examples/mean-reversion.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/alpacahq/alpaca-trade-api-go/alpaca"
+	"github.com/alpacahq/alpaca-trade-api-go/common"
+	"github.com/shopspring/decimal"
+)
+
+type alpacaClientContainer struct {
+	client         *alpaca.Client
+	closingPrices  []float64
+	runningAverage float64
+	lastOrder      string
+	timeToClose    float64
+	stock          string
+}
+
+var alpacaClient alpacaClientContainer
+
+func init() {
+	API_KEY := "API_KEY"
+	API_SECRET := "API_SECRET"
+	BASE_URL := "https://paper-api.alpaca.markets"
+
+	os.Setenv(common.EnvApiKeyID, API_KEY)
+	os.Setenv(common.EnvApiSecretKey, API_SECRET)
+	alpaca.SetBaseUrl(BASE_URL)
+
+	alpacaClient = alpacaClientContainer{
+		alpaca.NewClient(common.Credentials()),
+		[]float64{},
+		0.0,
+		"",
+		0.0,
+		"AAPL",
+	}
+}
+
+func main() {
+	// First, cancel any existing orders so they don't impact our buying power.
+	status, until, limit := "open", time.Now(), 100
+	orders, _ := alpacaClient.client.ListOrders(&status, &until, &limit)
+	for _, order := range orders {
+		_ = alpacaClient.client.CancelOrder(order.ID)
+	}
+
+	// Wait for market to open
+	fmt.Println("Waiting for market to open...")
+	for {
+		isOpen := alpacaClient.awaitMarketOpen()
+		if isOpen {
+			break
+		}
+		time.Sleep(2000 * time.Millisecond)
+	}
+	fmt.Println("Market Opened.")
+
+	// Get the running average of prices of the last 20 minutes, waiting until we have 20 bars from market open.
+	fmt.Println("Waiting for 20 bars...")
+	for {
+		layout := "2006-01-02T15:04:05.000Z"
+		rawTime, _ := time.Parse(layout, time.Now().String())
+		currTime := rawTime.String()
+		cal, _ := alpacaClient.client.GetCalendar(&currTime, &currTime)
+		marketOpen, _ := time.Parse(layout, cal[0].Open)
+		bars, _ := alpacaClient.client.GetSymbolBars(alpacaClient.stock, alpaca.ListBarParams{Timeframe: "minute", StartDt: &marketOpen})
+		if len(bars) >= 20 {
+			break
+		} else {
+			time.Sleep(60000 * time.Millisecond)
+		}
+	}
+	fmt.Println("We have 20 bars.")
+
+	amtBars := 20
+	bars, err := alpacaClient.client.GetSymbolBars(alpacaClient.stock, alpaca.ListBarParams{Timeframe: "minute", Limit: &amtBars})
+	if err != nil {
+		fmt.Println(err)
+	}
+	for _, bar := range bars {
+		alpacaClient.closingPrices = append(alpacaClient.closingPrices, float64(bar.Close))
+		if len(alpacaClient.closingPrices) > 20 {
+			alpacaClient.closingPrices = alpacaClient.closingPrices[1:]
+		}
+		alpacaClient.runningAverage = ((alpacaClient.runningAverage * float64(len(alpacaClient.closingPrices)-1)) + float64(bar.Close)) / float64(len(alpacaClient.closingPrices))
+	}
+
+	for {
+		alpacaClient.run()
+	}
+}
+
+// Rebalance our portfolio every minute based off running average data.
+func (alp alpacaClientContainer) run() {
+	if alpacaClient.lastOrder != "" {
+		_ = alp.client.CancelOrder(alpacaClient.lastOrder)
+	}
+
+	// Rebalance the portfolio.
+	alp.rebalance()
```

Comment:
Maybe do this in the else block below? Just so at the end of the day it doesn't rebalance and then immediately sell off.